### PR TITLE
Created cross server RPC stub doc

### PR DESCRIPTION
**Pull request by danielimprobable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

#### Primary reviewers
@girayimprobable  @joshuahuburn

### CODE REVIEW

File: SpatialGDK/Documentation/features.md [hunk 1]
```diff
@@ -50,9 +48,7 @@ Actor property handover is a new feature we’re introducing with the GDK. It re
 
 ## Cross-server RPCs
 
-`Cross-server RPC`s are a new type of RPC that functions between two servers. These enable a server without authority over an Actor to communicate with the server which has authority and is able to manipulate the Actor. For more information, see documentation.
-
-[//]: # (TODO: Add link to the Cross-server RPC doc)
+`Cross-server RPC`s are a new type of RPC that functions between two servers. These enable a server without authority over an Actor to communicate with the server which has authority and is able to manipulate the Actor. For more information see the documentation on [Cross-server RPCs]({{urlRoot}}/content/cross-server-rpcs).
```
[the remark this answers is not in the document]
Comment (danielimprobable) [reply SpatialGDK/Documentation/features.md]: I'll add a small comment there.

File: SpatialGDK/Documentation/content/cross-server-rpcs.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+## Cross-server RPCs
+
+### Overview
+
+In Unreal’s native client-server architecture, your game server holds the canonical state of the whole game world. As there is a single game server, it has authority over all server Actors and so it is able to invoke and execute functions on Actors unhindered. 
+
+In SpatialOS games, there can be more than one server; these multiple servers are known as “server-workers”. (Find out more about server-workers as well as “client-workers” in the [glossary]({{URLRoot}}/content/glossary#workers).)  
+As a SpatialOS game runs across many server-workers, SpatialOS server-workers have the concept of “worker authority” - where only one server-worker at a time is able to invoke and execute functions on Actors. (Find out more about authority in the [glossary]({{URLRoot}}/content/glossary#authority).)
```
Comment (danielimprobable) [inline SpatialGDK/Documentation/content/cross-server-rpcs.md]: The glossary links currently aren't present, @ElleEss could you confirm that these are valid please?

**Comment by danielimprobable** [comment]
@ElleEss @anne-edwards @girayimprobable @joshuahuburn @spatialos/develop-unreal this PR is now ready for review

File: SpatialGDK/Documentation/content/cross-server-rpcs.md (hunk 2 above)
Comment (anne-edwards) [reply SpatialGDK/Documentation/content/cross-server-rpcs.md]: I'll check all glossary links against the latest glossary GDoc.

File: SpatialGDK/Documentation/content/cross-server-rpcs.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+As Unreal expects there to be only one server, rather than several servers, the GDK has a custom solution to take advantage of the SpatialOS distributed server architecture. This involves handling the scenario where a server-worker attempts to invoke an RPC on an Actor that another server-worker has authority over. This custom solution is the cross-server RPC. The GDK offers cross-server RPC in addition to support for the [native RPC types that Unreal provides (Unreal documentation)](https://docs.unrealengine.com/en-us/Gameplay/Networking/Actors/RPCs).
+
+When a cross-server RPC is invoked by a non-authoritative server-worker, SpatialOS routes the execution through the SpatialOS [runtime]({{URLRoot}}/contents/glossary#runtime) to the authoritative server-worker - this authoritative server-worker executes the RPC.
```
Comment (anne-edwards) [inline SpatialGDK/Documentation/content/cross-server-rpcs.md]: glossary#runtime -> glossary#spatialos-runtime

Comment (anne-edwards) [inline SpatialGDK/Documentation/content/cross-server-rpcs.md]: runtime -> Runtime

File: SpatialGDK/Documentation/content/cross-server-rpcs.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+The example diagram below shows a player successfully shooting another player’s hat across a server-worker boundary.
+
+![A situation where you might need cross-server RPCs]({{assetRoot}}assets/shooting-workflow-simple.png)
+
+In the diagram, Server-worker 1 has authority over Player 1 and Server-worker 2 has authority over Player 2. If Player 1 shoots a bullet, Server-worker 1 knows about the bullet and can make any necessary changes to Player 1 but it can’t make changes to Player 2 when the bullet hits. SpatialOS ensures that Server-worker 2 can affect changes to Player 2 (the hat gets hit by the bullet) by routing the change notification from Server-worker 1 to Server-worker 2.
```
Comment (anne-edwards) [inline SpatialGDK/Documentation/content/cross-server-rpcs.md]: affect -> make

File: SpatialGDK/Documentation/content/cross-server-rpcs.md (hunk 3 above)
Comment (danielimprobable) [reply SpatialGDK/Documentation/content/cross-server-rpcs.md]: @anne-edwards Are you sure that it should be capital R in the link? I thought the improbadoc links always were lower case?

Comment (anne-edwards) [reply SpatialGDK/Documentation/content/cross-server-rpcs.md]: The instance of the word "Runtime" in square brackets should have a capital (because that's how we talk about "the SpatialOS Runtime", partly to distinguish it from something happening "at runtime"). But you're right, the word in the link itself should stay as it is :)

Comment (danielimprobable) [reply SpatialGDK/Documentation/content/cross-server-rpcs.md]: Great, sorry misunderstood you updating now:)

File: SpatialGDK/Documentation/content/cross-server-rpcs.md (hunk 2 above)
Comment (ElleEss) [reply SpatialGDK/Documentation/content/cross-server-rpcs.md]: Thank you <3 - assuming I don't need to check as Anne's all over it?

Comment (ElleEss) [reply SpatialGDK/Documentation/content/cross-server-rpcs.md]: ALso, note that linting will catch this before we go live.

File: SpatialGDK/Documentation/features.md [hunk 5]
```diff
@@ -50,9 +50,7 @@ Actor property handover is a new feature we’re introducing with the GDK. It re
...
```
Comment (ElleEss) [inline SpatialGDK/Documentation/features.md]: glossary link to authority. (Put this in backticks so you can see the actual link)
`[authority]({{urlRoot}}/content/glossary#authority)`

Comment (ElleEss) [reply SpatialGDK/Documentation/features.md]: "These enable a server without [authority]({{urlRoot}}/content/glossary#authority) over an Actor to communicate with the server which has authority and is able to manipulate the Actor."

-> The sentence is hard to follow if you don't already know what this means. It assumes that you know you need worker authority to work in a multiserver environment and one of the problems of worker authority is cross-server communication. They may know this but we can probably explain it more outside in.
I think the key is; any Actor can affect a change over any other Actors, regardless of which server has authority over them.

Suggest...

"As SpatialOS has multiple servers, only one server has the [authority]({{urlRoot}}/content/glossary#authority) to change an Actor at any one time. However, as actions by Actors under the authority of different servers need to affect change over each other, cross-server RPC allows this, regardless of what server has authority over an Actor."

Comment (ElleEss) [reply SpatialGDK/Documentation/features.md]: ?? Not sure if it's clearer. :-)

Comment (ElleEss) [reply SpatialGDK/Documentation/features.md]: As SpatialOS has multiple servers, only one server has the [authority]({{urlRoot}}/content/glossary#authority) to manipulate an Actor at any one time. Cross-server RPCs enable a server without authority over an Actor to tell the server which has authority over that Actor to manipulate it. For more information see the documentation on [Cross-server RPCs]({{urlRoot}}/content/cross-server-rpcs).

File: SpatialGDK/Documentation/content/cross-server-rpcs.md (hunk 2 above)
Comment (anne-edwards) [inline SpatialGDK/Documentation/content/cross-server-rpcs.md]: Just had a quick skim through the doc again and I spotted that this line (#8) either needs to become part of line 7, or have an extra blank line before it.

image

File: SpatialGDK/Documentation/content/cross-server-rpcs.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        // Implementation goes here...
+    }
+   ```
+   Note: You may need to implement the `MyCrossServerRPC_Validation()` if you used the `WithValidation` attribute.
+
+1. Invoke the `CrossServer` RPC function as you would with any other function.
+
+### How to send a cross-server RPC (using Blueprints)
+
+To set up a cross-server RPC in a Blueprint, follow the same instructions as you would for marking up RPCs in Blueprints (Unreal documentation), but from the **Replicates** drop-down list within the **Details** panel of your event, select **Run on authoritative server (sent from server)**:
```
Comment (anne-edwards) [inline SpatialGDK/Documentation/content/cross-server-rpcs.md]: `marking up RPCs in Blueprints (Unreal documentation)`

is missing a link

Comment (anne-edwards) [reply SpatialGDK/Documentation/content/cross-server-rpcs.md]: looks like this is the link you need (from the GDoc):

https://docs.unrealengine.com/en-us/Gameplay/Networking/Actors/RPCs#blueprints

File: SpatialGDK/Documentation/features.md [hunk 7]
```diff
@@ -50,9 +50,9 @@ Actor property handover is a new feature we’re introducing with the GDK. It re
 
 ## Cross-server RPCs
 
-`Cross-server RPC`s are a new type of RPC that functions between two servers. These enable a server without authority over an Actor to communicate with the server which has authority and is able to manipulate the Actor. For more information, see documentation.
+As SpatialOS has multiple servers, only one server has the authority to manipulate an Actor at any one time. Cross-server RPCs enable a server without [authority]({{urlRoot}}/content/glossary#authority) over an Actor to tell the server which has authority over that Actor to manipulate it. For more information see the documentation on [Cross-server RPCs]({{urlRoot}}/content/cross-server-rpcs).
```
[the remark this answers is not in the document]
Comment (danielimprobable) [reply SpatialGDK/Documentation/features.md]: I'll touch it up now:)
